CJC-416

ccaprio97 opened the pull request:
## What’s changing?

Changing status in upload table to 'Complete' on successful upload

## Why?

NAC-416

## This PR has:

- [ ] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [ ] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [ ] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [ ] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)

## Anything else?

Unaware of any documentation to update because all affected code is internal. If some needs to be added where would that go?

bgfoshay reviewed (changes requested):
I'm not sure this is the right approach. I think we need to move the initial record creation outside the transaction. And it should still set to "Uploading". Then we'll need a new statement, at the end of the transaction (after participants are finished being copied over) to update the upload record's status to "Complete" (or "Failed" when appropriate). Call me if you want to discuss.

ccaprio97 commented:
changed the original place back to 'uploading' and added new method after addParticipants. Returns the Id even though it doesnt need to. will probably need to have the id being returned when we eventually check for failed or successful plus it doesnt hurt

bgfoshay commented on participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs:
```diff
@@ -63,6 +63,7 @@ public async Task AddParticipants(IEnumerable<IParticipant> participants,string
                 });
 
                 await _participantDao.AddParticipants(participantDbos);
+                await _uploadDao.UpdateUploadStatus(upload);
```
Issue: I think you also need to move the following line outside/above the transaction scope.

var upload = await _uploadDao.AddUpload(uploadIdentifier);

bgfoshay commented on participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs:
```diff
@@ -50,20 +51,27 @@ public async Task AddParticipants(IEnumerable<IParticipant> participants,string
+                }
+            }
+            catch (Exception ex) {
+                await _uploadDao.UpdateUploadStatus(upload, UploadStatuses.Failed.ToString());
```
Thought: You should log the error here so that if one occurs, we can see what happened.

bgfoshay replied:
I think it should be a simple change and if so, we should log it as part of this PR. Unless you have concerns with doing so?

bgfoshay commented on participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs:
```diff
@@ -50,20 +51,28 @@ public async Task AddParticipants(IEnumerable<IParticipant> participants,string
-                var upload = await _uploadDao.AddUpload(uploadIdentifier);
-
-                var participantDbos = participants.Select(p => new ParticipantDbo(p)
+            var upload = await _uploadDao.AddUpload(uploadIdentifier);
```
Thought: Do we have a unit test case for upload record created but participants failed?

ccaprio97 replied:
check AfterException_AddParticipantsRollsTranactionBack

bgfoshay commented on participants/tests/Piipan.Participants.Core.IntegrationTests/ParticipantServiceTests.cs:
```diff
@@ -84,7 +84,7 @@ public async void AfterException_AddParticipantsRollsTranactionBack(int nPartici
                 {
                     long expectedNewUploadId = ++lastUploadId;
-                    long actualLastUploadId = GetLastUploadId();
+                    long actualLastUploadId = GetLastSuccessfulUploadId();
```
Thought: Aside from verifying the most recent successful uploadId, I think you should verify that the latest upload has a status = FAILED.

bgfoshay commented on shared/src/Piipan.Shared.TestFixtures/ParticipantsDbFixture.cs:
```diff
@@ -129,5 +129,35 @@ public Int64 GetLastUploadId()
+            return result;
+        }
+
+        public Int64 GetLastFailedUploadId()
```
Question: can you change/consolidate GetLastSuccessfulUploadId/GetLastFailedUploadId -> GetLatestUploadByStatus(string desiredStatus)?

bgfoshay reviewed (approved):
👍Thanks for addressing all of my comments!
